### DISCUSSION TITLE
How to include the groups claim in an access_token (JWT)

### DISCUSSION THREAD

**Discussion by tsschaffert** [authelia/authelia#8073] [Q&A] [answered]
Is it possible to configure Authelia to include the `groups` claim in a JWT-formatted access token?

We configured our client to receive the access token as a JWT (via [access_token_signed_response_alg](https://www.authelia.com/configuration/identity-providers/openid-connect/clients/#access_token_signed_response_alg)). But in the resulting JWT, there is no `groups` claim. We have provided the required scope, so when calling the `/userinfo` route, we get all the information we need, but it would be great if there was a way to get this information directly from the JWT.

**Comment by james-d-elliott** [comment, marked as answer]
So while there is provision for identity claims in these access tokens for interoperability, the specification does not give any guidance on the way in which this is to be determined. We are aiming for specification compliance and are generally going to avoid unofficial or unspecified practices where possible. 

In addition it's mentioned in [the spec](https://datatracker.ietf.org/doc/html/rfc9068#PrivacyConsiderations) is the following:

> The client MUST NOT inspect the content of the access token: the authorization server and the resource server might decide to change the token format at any time (for example, by switching from this profile to opaque tokens); hence, any logic in the client relying on the ability to read the access token content would break without recourse. The OAuth 2.0 framework assumes that access tokens are treated as opaque by clients. Administrators of authorization servers should also take into account that the content of an access token is visible to the client. Whenever client access to the access token content presents privacy issues for a given scenario, the authorization server needs to take explicit steps to prevent them.

This means the specification does not form this kind of contract between the authorization server and clients. The access token is only intended to be understood by the authorization server or related resource servers to access permitted resources.

The ID Token forms the contract by which identity information specifically the proof of identity via the `sub` and `iss` claims and the hash of the access token in order to access identity information. Clients can use this access token to perform a query against the User Info endpoint for the accessible identity information. This ID Token as its intended purpose (see [Requesting Claims using the "claims" Request Parameter](https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter)). This parameter is currently not supported by Authelia however will be in 4.39, and until such a time the ID Token returns all claims that the granted scopes have access to (again, changing in 4.39).

In contrast to the Access Token the ID Token and User Info endpoint is meant to be inspected by the client, and expressly intended for this purpose.

**Reply by tsschaffert** [reply]
That does make a lot of sense, thank you for the detailed explanation.
